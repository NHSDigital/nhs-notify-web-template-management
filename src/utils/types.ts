export type FormId =
  | 'choose-a-template-type'
  | 'create-nhs-app-template'
  | 'create-email-template'
  | 'create-letter-template'
  | 'create-text-message-template'
  | 'preview-nhs-app-template'
  | 'submit-template'
  | 'create-nhs-app-template-back';

export type FormErrorState = {
  formErrors: string[];
  fieldErrors: Record<string, string[]>;
};

export enum TemplateType {
  NHS_APP = 'NHS_APP',
  SMS = 'SMS',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
}

export type Session = {
  id: string;
  templateType: TemplateType | 'UNKNOWN';
  nhsAppTemplateName: string;
  nhsAppTemplateMessage: string;
};

export type FormState = {
  validationError?: FormErrorState;
  redirect?: string;
};

export type TemplateFormState = FormState & Session;

export type PageProps = {
  params: {
    sessionId: string;
  };
};

export type TemplateSubmittedPageProps = {
  params: {
    templateId: string;
  };
};

export type PageComponentProps = {
  initialState: TemplateFormState;
};

export type SubmitTemplatePageComponentProps = {
  templateName: string;
  sessionId: string;
};

export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
