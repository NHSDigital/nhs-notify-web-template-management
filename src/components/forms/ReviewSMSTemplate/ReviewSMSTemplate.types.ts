export type ReviewSMSTemplateProps = {
  templateName: string;
  message: string;
};
