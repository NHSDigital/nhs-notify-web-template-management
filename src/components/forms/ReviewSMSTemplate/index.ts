export * from './ReviewSMSTemplate';
export * from './ReviewSMSTemplate.types';
export * from './server-actions';
