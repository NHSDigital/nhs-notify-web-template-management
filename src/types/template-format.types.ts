export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
