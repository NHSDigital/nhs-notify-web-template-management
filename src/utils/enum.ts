export enum TemplateType {
  NHS_APP = 'NHS_APP',
  SMS = 'SMS',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
}
