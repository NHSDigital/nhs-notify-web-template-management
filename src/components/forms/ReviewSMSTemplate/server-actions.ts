import { MarkdownItWrapper } from '@utils/markdownit';
import { TemplateFormState } from '@utils/types';
import { redirect, RedirectType } from 'next/navigation';
import { z } from 'zod';

export const $FormSchema = z.object({
  reviewSMSTemplateAction: z.enum(['sms-edit', 'sms-submit'], {
    message: 'Select an option',
  }),
});

export function renderMarkdown(
  value: string,
  markdown = new MarkdownItWrapper()
) {
  return markdown.render(value);
}

export async function reviewSmsTemplateAction(
  formState: TemplateFormState,
  formData: FormData
) {
  const form = Object.fromEntries(formData.entries());

  const { success, error, data } = $FormSchema.safeParse(form);

  if (!success) {
    return {
      ...formState,
      validationError: error.flatten(),
    };
  }

  if (data.reviewSMSTemplateAction === 'sms-edit') {
    return redirect(
      `/create-text-message-template/${formState.id}`,
      RedirectType.push
    );
  }

  if (data.reviewSMSTemplateAction === 'sms-submit') {
    return redirect(
      `/submit-text-message-template/${formState.id}`,
      RedirectType.push
    );
  }

  throw new Error('Unknown review sms template action.');
}
