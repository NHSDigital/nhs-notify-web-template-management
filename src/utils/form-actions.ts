'use server';

import { getAmplifyBackendClient } from '@utils/amplify-utils';
import { DbOperationError } from '@domain/errors';
import { Template, Draft } from './types';
import { logger } from './logger';

export async function createTemplate(
  template: Draft<Template>
): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.create(template);

  if (errors || !data) {
    logger.error('Failed to create template', errors);
    throw new Error('Failed to create new template');
  }
  return data;
}

export async function saveTemplate(template: Template): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.update(template);

  if (errors) {
    logger.error('Failed to save template', errors);
    throw new Error('Failed to save template data');
  }

  if (!data) {
    throw new DbOperationError({
      message:
        'Template in unknown state. No errors reported but entity returned as falsy',
      operation: 'update',
    });
  }
  return data;
}

export async function getTemplate(
  templateId: string
): Promise<Template | undefined> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.get({
      id: templateId,
    });

  if (errors) {
    logger.error('Failed to get template', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve template for ID ${templateId}`);
    return undefined;
  }

  return data;
}

export async function sendEmail(
  templateId: string,
  templateName: string,
  templateMessage: string,
  templateSubjectLine: string | null
) {
  const res = await getAmplifyBackendClient().queries.sendEmail({
    recipientEmail: 'england.test.cm@nhs.net',
    templateId,
    templateName,
    templateMessage,
    templateSubjectLine,
  });

  if (res.errors) {
    logger.error({
      description: 'Error sending email',
      res,
    });
  }
}

export async function getTemplates(): Promise<Template[] | []> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.list();

  if (errors) {
    logger.error('Failed to get templates', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve templates`);
    return [];
  }

  return data;
}
