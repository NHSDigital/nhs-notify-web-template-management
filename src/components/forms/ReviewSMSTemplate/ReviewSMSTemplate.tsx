'use client';

import Link from 'next/link';
import { PreviewTemplate } from '@molecules/PreviewTemplate';
import { ReviewTemplate } from '@organisms/ReviewTemplate';
import content from '@content/content';
import { PageComponentProps, SMSTemplate } from '@utils/types';
import { useFormState } from 'react-dom';
import { BackLink } from 'nhsuk-react-components';
import { getBasePath } from '@utils/get-base-path';
import { renderMarkdown, reviewSmsTemplateAction } from './server-actions';

export function ReviewSMSTemplate({
  initialState,
}: Readonly<PageComponentProps<SMSTemplate>>) {
  const {
    components: {
      reviewSMSTemplateContent: { sectionHeading, form },
    },
  } = content;

  const [state, action] = useFormState(reviewSmsTemplateAction, initialState);

  const templateMessage = initialState.message;

  const html = renderMarkdown(templateMessage);

  return (
    <div className='nhsuk-grid-row'>
      <BackLink
        href={`${getBasePath()}/manage-templates`}
        className='nhsuk-u-margin-bottom-5 nhsuk-u-margin-left-3'
      >
        Back to all templates
      </BackLink>
      <ReviewTemplate
        template={initialState}
        sectionHeading={sectionHeading}
        form={{
          ...form,
          state,
          action,
          formId: 'review-sms-template',
          radiosId: 'reviewSMSTemplateAction',
        }}
        PreviewComponent={<PreviewTemplate.Sms message={html} />}
      />
      <p>
        <Link href='/manage-templates'>Back to all templates</Link>
      </p>
    </div>
  );
}
