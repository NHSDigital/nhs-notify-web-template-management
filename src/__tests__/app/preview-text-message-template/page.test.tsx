import PreviewSMSTemplatePage from '@app/preview-text-message-template/[templateId]/page';
import { ReviewSMSTemplate } from '@forms/ReviewSMSTemplate';
import { SMSTemplate } from '@utils/types';
import { TemplateType } from '@utils/enum';
import { redirect } from 'next/navigation';
import { getTemplate } from '@utils/form-actions';

jest.mock('@utils/form-actions');
jest.mock('next/navigation');
jest.mock('@forms/ReviewSMSTemplate');
jest.mock('@utils/logger');

const redirectMock = jest.mocked(redirect);
const getTemplateMock = jest.mocked(getTemplate);

describe('PreviewSMSTemplatePage', () => {
  beforeEach(jest.resetAllMocks);

  it('should load page', async () => {
    const state: SMSTemplate = {
      id: 'template-id',
      version: 1,
      templateType: TemplateType.SMS,
      SMS: {
        name: 'template-name',
        message: 'template-message',
      },
    };

    getTemplateMock.mockResolvedValueOnce(state);

    const page = await PreviewSMSTemplatePage({
      params: {
        templateId: 'template-id',
      },
    });

    expect(page).toEqual(<ReviewSMSTemplate initialState={state} />);
  });

  it('should redirect to invalid-template when no template is found', async () => {
    await PreviewSMSTemplatePage({
      params: {
        templateId: 'template-id',
      },
    });

    expect(redirectMock).toHaveBeenCalledWith('/invalid-template', 'replace');
  });

  test.each([
    {
      templateType: TemplateType.LETTER,
      SMS: {
        name: 'template-name',
        message: 'template-message',
      },
    },
    {
      templateType: TemplateType.EMAIL,
      SMS: {
        name: 'template-name',
        message: 'template-message',
      },
    },
    {
      templateType: TemplateType.NHS_APP,
      SMS: {
        name: 'template-name',
        message: 'template-message',
      },
    },
    {
      templateType: TemplateType.SMS,
      SMS: {
        name: 'template-name',
        message: undefined as unknown as string,
      },
    },
    {
      templateType: TemplateType.SMS,
      SMS: {
        name: undefined as unknown as string,
        message: 'template-message',
      },
    },
    {
      templateType: TemplateType.SMS,
      SMS: {
        name: null as unknown as string,
        message: null as unknown as string,
      },
    },
  ])(
    'should redirect to invalid-template when template is $templateType and name is $smsTemplateName and message is $smsTemplateMessage',
    async (value) => {
      getTemplateMock.mockResolvedValueOnce({
        id: 'template-id',
        version: 1,
        ...value,
      });

      await PreviewSMSTemplatePage({
        params: {
          templateId: 'template-id',
        },
      });

      expect(redirectMock).toHaveBeenCalledWith('/invalid-template', 'replace');
    }
  );
});
