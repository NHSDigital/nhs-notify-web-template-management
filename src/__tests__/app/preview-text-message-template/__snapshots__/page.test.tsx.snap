// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewSMSTemplatePage should load page 1`] = `
<ReviewSMSTemplate
  initialState={
    {
      "id": "session-id",
      "nhsAppTemplateMessage": "",
      "nhsAppTemplateName": "",
      "smsTemplateMessage": "template-message",
      "smsTemplateName": "template-name",
      "templateType": "SMS",
    }
  }
/>
`;
